### DISCUSSION TITLE
Virtual Scrolling

### DISCUSSION THREAD

**Discussion by knpwrs** [solidjs/solid#512] [Q&A] [answered]
Does the Solid ecosystem have any libraries to support virtual scrolling (ala [`react-cool-virtual`](https://github.com/wellyshen/react-cool-virtual), [`react-window`](https://github.com/bvaughn/react-window), or [`react-virtualized`](https://github.com/bvaughn/react-virtualized))? If not, are there any generic virtual scrolling libraries that work well with Solid?

**Comment by mosheduminer** [comment, marked as answer]
I don't think the ecosystem has such libraries yet, but there are some codesandboxes that are basically complete implementations (these were shared a while back on the solid discord by @rturnq ).
https://codesandbox.io/embed/virtualized-list-9dw9q
https://codesandbox.io/embed/virtualized-sheet-8fokq
https://codesandbox.io/embed/virtualized-table-ks63n

When using these, keep in mind that imports of `createState` from "solid-js" need to be changed to `createStore` from "solid-js/store" in order to work with the 1.0 release.

**Comment by floer32** [comment]
More recently there's also: https://github.com/minht11/solid-virtual-container (note: I haven't tried it in a production application yet)

**Comment by jchook** [comment]
[@tanstack/solid-virtual](https://tanstack.com/virtual/v3) looks promising, but still in beta and currently not yet published to NPM.

You could still use the UI-agnostic core and adapt the [WIP Solid implementation](https://github.com/TanStack/virtual/blob/beta/packages/solid-virtual/src/index.tsx)

**Comment by knpwrs** [comment]
I've been thinking that a generic virtualization engine that could connect to multiple rendering engines would be cool. Not sure how practical that is. Something like https://github.com/wellyshen/react-cool-virtual just not bound to any particular framework.

**Comment by chitoadinugraha** [comment]
I tried to make virtual scrolling usign raw javascript, it support dynamic sized item. 

Here is my code
```
<!DOCTYPE html>

<html>

<head>
    <style>
        #container {
            position: absolute;
            top: 180px;
            left: 80px;
            width: 200px;
            height: 300px;

            border: 1px solid #888;
            overflow: visible;
        }

        #inner {
            position: absolute;
            left: 0;
            right: 0;
            border: 1px solid red;
        }
    </style>
</head>

<body>
    <div id="container">
        <div id="inner"></div>
    </div>


    <script>
        const container = document.getElementById('container');
        const inner = document.getElementById('inner');

        const overscan = 100;
        let originalY = 0;
        let startY = 0;

        let firstChildIdx = 0;
        let lastChildIdx = 0;

        const getItem = (idx) => {
            const div = document.createElement('div');
            div.innerText = 'Item ' + idx;
            return div;
        }

        const updateOffset = () => {
            inner.style.top = `${startY}px`;
        }

        const scrollDown = (dY) => {
            startY -= dY;
            originalY -= dY;

            const fillAfter = () => {
                const containerHeight = container.clientHeight;

                while (startY + inner.clientHeight < containerHeight + overscan) {
                    const item = getItem(lastChildIdx++);
                    inner.append(item);
                }
            }

            const trimBefore = () => {

                while (startY < -overscan) {
                    const firstChild = inner.firstChild;

                    // advance sY
                    startY += firstChild.clientHeight;
                    firstChild.remove();
                    firstChildIdx++
                }

            }

            fillAfter();
            trimBefore();
            updateOffset();
        }

        scrollDown(0);

        const scrollUp = (dY) => {
            console.log('scroll up');
            const containerHeight = container.clientHeight;

            const fillBefore = () => {

                while (startY > -overscan && firstChildIdx > 0) {

                    const idx = --firstChildIdx;
                    const item = getItem(idx);
                    inner.insertBefore(item, inner.firstChild);
                    startY -= item.clientHeight;
                }
            }

            const trimAfter = () => {
                while (startY + inner.clientHeight > containerHeight + overscan) {
                    const lastChild = inner.lastChild;

                    lastChild.remove();
                    lastChildIdx--;
                    // if (lastChildIdx < 0) lastChildIdx = 0;
                }
            }

            if (originalY + dY > 0) {
                startY = 0;
                originalY = 0;

            } else {

                originalY += dY
                startY += dY;

                fillBefore();
            }

            trimAfter();
            updateOffset();
        }

        container.onwheel = (e) => {
            const dY = e.deltaY;

            if (dY > 0) {
                scrollDown(Math.abs(dY))
            } else {
                scrollUp(Math.abs(dY));
            }
        }


    </script>
</body>

</html>
```

**Reply by Grabber** [reply]
It doesn't work if you drag the scrollbars nor if you press the up/down or page up/down keys.

If you use the 'scroll' event it works, but on Firefox you will get a warning about the scroll-linked effect.

**Reply by Grabber** [reply]
I did some tests that 'onwheel' could be replaced with `addEventListener('scroll', ...)` and `removeEventListener('scroll', ...)`. Next, call window.requestAnimationFrame(() => { /* update... */})` inside the scroll handler function. It solves the Firefox scroll-linked performance issue and also the console warning!

Most of the virtualized list libraries are not doing it nowadays!

**Comment by inokawa** [comment]
I made a virtualization library supporting dynamic sized items. Its Solid support is newborn but the shared logic is almost mature.
https://github.com/inokawa/virtua

**Reply by onx2** [reply]
This is amazing, thank you for releasing!

**Comment by Brendonovich** [comment]
It's been mentioned here already but Tanstack's Solid Virtual has been out for a while and has worked great for us.

https://tanstack.com/virtual
https://github.com/tanstack/virtual

**Comment by chitoadinugraha** [comment]
https://solidjs-use.github.io/solidjs-use/core/useVirtualList
